### DISCUSSION TITLE
Known issues with the API

### DISCUSSION THREAD

**Discussion by bhrutledge** [bhrutledge/ma-legislature-api#1] [General]
This is a list of issues that I discovered while exploring the API (via [Jupyter notebooks](https://github.com/bhrutledge/ma-legislature-api/tree/main/notebooks)) and making the [bill details map](https://bhrutledge.com/ma-legislature-api/pages/bill-details.html). It might be worth breaking them out into discrete discussion topics.

---

- Most (if not all) of the URLs in the responses are `http` instead of `https`
- Responses don't include a URL for viewing an object on the website
    - e.g. <https://malegislature.gov/Legislators/Profile/E_U1>, <https://malegislature.gov/Bills/192/H805>
    - Can be constructed from response fields, but maybe add `URL` for convenience
- The Swagger docs use `camelCase` for the example responses, while the actual responses use `PascalCase`
- [`/LegislativeMembers/{memberCode}`](https://malegislature.gov/api/swagger/index.html?url=/api/swagger/v1/swagger.json#/LegislativeMembers/LegislativeMembers_GetLegislativeMembers3)
    - Takes a long time to respond
    - Includes all `SponsoredBills`, `CoSponsoredBills`, and `Committees`
        - Maybe use query parameters to include/exclude them (e.g. `?includeBills=0&includeCommittees=0`)
        - Also applies to other endpoints
    - Doesn't include first/last name (shown on the [member list page](https://malegislature.gov/Legislators/Members/Senate))
    - Doesn't include title (i.e. "Senator", "Representative")
- [`/Committees/{committeeCode}`](https://malegislature.gov/api/swagger/index.html?url=/api/swagger/v1/swagger.json#/Committees/Committees_GetCommittees3)
    - Includes all `DocumentsBeforeCommittee` and `ReportedOutDocuments`
        - Maybe use a query parameter to include/exclude them
    - Doesn't include the committee members (shown on the [committee detail page](https://malegislature.gov/Committees/Detail/J19/192))
        - Finding that requires searching all of the `LegislativeMembers.Committees`
        - Maybe add `HouseMembers` and `SenateMembers` via the `LegislativeMemberSummary` model
    - Doesn't include the Vice Chair
- [`/Documents/{documentNumber}`](https://malegislature.gov/api/swagger/index.html?url=/api/swagger/v1/swagger.json#/Documents/Documents_GetDocument2)
    - Doesn't include the relevant committees
        - Finding that requires searching all of the `Committees.DocumentsBeforeCommittee` and `Committees.ReportedOutDocuments`
        - `CommitteeRecommendations` includes the committee, but not clear if that covers all `Committees.ReportedOutDocuments`
        - Maybe add `BeforeCommittee` and `ReportedOut` via the `CommitteeSummary` model
    - `PrimarySponsor` and `Cosponsors` use the `BillSponsorSummary` model, which doesn't include `GeneralCourtNumber` or `Branch`, like the `LegislativeMemberSummary` model does
- [`RollCalls`](https://malegislature.gov/api/swagger/index.html?url=/api/swagger/v1/swagger.json#/RollCalls)
    - No endpoint to get all roll calls for a given `generalCourtNumber` and `branch`
- [`/RollCalls/{rollCallNumber}`](https://malegislature.gov/api/swagger/index.html?url=/api/swagger/v1/swagger.json#/RollCalls/RollCalls_GetRollCall2)
    - Doesn't include the relevant `Documents` (e.g. via the `DocumentSummary` model)
- [`/Hearings`](https://malegislature.gov/api/swagger/index.html?url=/api/swagger/v1/swagger.json#/Hearings/Hearings_GetHearingsAll)
    - Includes all-time hearings, instead of just the current general court
- [`/Hearings/{eventId}`](https://malegislature.gov/api/swagger/index.html?url=/api/swagger/v1/swagger.json#/Hearings/Hearings_GetHearings)
    - Has lots of 500 responses

**Comment by laurenmarietta** [comment]
Adding on my observations, it seems as though the following pages are simply broken:
- [/Branches/{branch}/Leadership](https://malegislature.gov/api/Branches/Senate/Leadership)
- [/GeneralCourts/{generalCourtNumber}/Branches/{branch}/Leadership](https://malegislature.gov/api/GeneralCourts/192/Branches/Senate/Leadership)
- [/GeneralCourts/{generalCourtNumber}/HouseJournals](https://malegislature.gov/api/GeneralCourts/192/HouseJournals)
- [/GeneralCourts/{generalCourtNumber}/SenateJournals](https://malegislature.gov/api/GeneralCourts/192/SenateJournals)

I also suspect that the roll call API is broken for House bills. I have no trouble [listing all roll calls](https://malegislature.gov/api/GeneralCourts/192/Documents/S10) or the [votes for a given roll call](https://malegislature.gov/api/GeneralCourts/192/Branches/Senate/RollCalls/2) for senate bill S.10. However for House bill H.68, which I know [has seen votes](https://malegislature.gov/Journal/House/192/RollCalls/hj02242021_RollCalls.pdf), the API [comes up blank](https://malegislature.gov/api/GeneralCourts/192/Documents/H68/RollCalls).

**Comment by jessemortenson** [comment]
I did some exploration of documents reported back at the session level, eg `https://malegislature.gov/api/GeneralCourts/192/Documents` then make a second request for each document in the result to `https://malegislature.gov/api/GeneralCourts/192/Documents/{documentID}`  Some observations on bills that are incomplete or return error codes:

- 114 have a `null` value for the `Details` property, so the second request can't be made. None of these appear to show up on the MA General Court bill search UI currently
- 24 return a `400` error `The requested Document could not be found...`. Similarly, none of these appear to show up in the MA General Court bill search UI currently.
- 54 return a `500` error with `{ "Message": "An error has occurred." }` JSON response. All of these appear to be "Report" type documents, not legislation in the sense of changing the law. Some examples:
  - https://malegislature.gov/Bills/192/HD4322
  - https://malegislature.gov/Bills/192/HD4135
  - https://malegislature.gov/Bills/192/HD4148
  - https://malegislature.gov/Bills/192/SD2675

**Reply by bhrutledge** [reply]
Hi @jessemortenson! What brings you here?

**Reply by jessemortenson** [reply]
I work for Civic Eagle. We adopted Open States into the company last year, and so found a link here from your comments on an Open States issue. Appreciate your work!
